memory/src/main/scala/filodb.memory/format/NibblePack.scala

```diff
@@ -0,0 +1,283 @@
+    bufpos += 1
+
+    if (bitmask != 0) {
+      // figure out min # of nibbles to represent nonzero words
+      var minLeadingZeros = 64
+      var minTrailingZeros = 64
+      for { i <- 0 until 8 optimized } {
+        minLeadingZeros = Math.min(minLeadingZeros, java.lang.Long.numberOfLeadingZeros(input(i)))
+        minTrailingZeros = Math.min(minTrailingZeros, java.lang.Long.numberOfTrailingZeros(input(i)))
+      }
+
+      val trailingNibbles = minTrailingZeros / 4
+      val numNibbles = 16 - (minLeadingZeros / 4) - trailingNibbles
+      val nibbleWord = (((numNibbles - 1) << 4) | trailingNibbles)
+      buf.putByte(bufpos, nibbleWord.toByte)
+      bufpos += 1
+
+      // Decide which packer to use
+      bufpos = packUniversal(input, buf, bufpos, numNibbles, trailingNibbles)
+    }
+
+    bufpos
+  }
+
+  // Returns the final bufpos
+  private def packUniversal(inputs: Array[Long], buf: MutableDirectBuffer, bufindex: Int,
+                    numNibbles: Int, trailingZeroNibbles: Int): Int = {
+    var bufpos = bufindex
+    val trailingShift = trailingZeroNibbles * 4
+    val numBits = numNibbles * 4
+    var outWord = 0L
+    var bitCursor = 0
+
+    for { i <- 0 until 8 optimized } {
+      val input = inputs(i)
+      if (input != 0) {
+        val remaining = 64 - bitCursor
+        val shiftedInput = input >> trailingShift
+
+        // This is least significant portion of input
+        outWord |= shiftedInput << bitCursor
+
+        // Write out current word if we've used up all 64 bits
+        if (remaining <= numBits) {
+          buf.putLong(bufpos, outWord, LITTLE_ENDIAN)
+          bufpos += 8
+          if (remaining < numBits) {
+              // Most significant portion left over from previous word
+              outWord = shiftedInput >> remaining
+          } else {
+              outWord = 0    // reset for 64-bit input case
+          }
+        }
+
+        bitCursor = (bitCursor + numBits) % 64
+      }
+    }
+
+    // Write remainder word if there are any bits remaining, and only advance buffer right # of bytes
+    if (bitCursor > 0) {
+        buf.putLong(bufpos, outWord, LITTLE_ENDIAN)
+        bufpos += ((bitCursor + 7) / 8)
+    }
+
+    bufpos
+  }
+
+  private val tlTempArray = new ThreadLocal[Array[Long]]()
+  def tempArray: Array[Long] = tlTempArray.get match {
+    case UnsafeUtils.ZeroPointer => val newArray = new Array[Long](8)
+                                    tlTempArray.set(newArray)
+                                    newArray
+    case a: Array[Long]          => a
+  }
+
+  sealed trait UnpackResult
+  case object Ok extends UnpackResult
+  case object InputTooShort extends UnpackResult
+
+  val empty = Array.empty[Byte]
+
+  trait Sink {
+    def process(data: Long): Unit
+  }
+
+  final case class DeltaSink(outArray: Array[Long]) extends Sink {
+    private var current: Long = 0L
+    private var pos: Int = 0
+    def process(data: Long): Unit = {
+      if (pos < outArray.size) {
```

Comment:
Why check and throw data away? Just remove the check and have an exception be thrown instead?

Reply:
I remember why this is necessary now.  It's because NibblePack always works in multiples of 8.  So, if a user packed 12 values, they will pass in an output array of 12 values and expect it to work, but in reality NibblePack demands 16.   I will add a comment.